### PR TITLE
Spawn a future to start the collator

### PR DESCRIPTION
**Pull request by cecton**
Related to https://github.com/paritytech/cumulus/issues/44

### CODE REVIEW

File: collator/src/lib.rs [hunk 1]
```diff
@@ -461,51 +512,52 @@ fn compute_targets(para_id: ParaId, session_keys: &[ValidatorId], roster: DutyRo
...
+pub fn run_collator<P, B>(
+	build_parachain: B,
 	para_id: ParaId,
 	key: Arc<CollatorPair>,
 	config: Configuration,
 ) -> polkadot_cli::error::Result<()> where
 	P: BuildParachainContext,
 	P::ParachainContext: Send + 'static,
 	<P::ParachainContext as ParachainContext>::ProduceCandidate: Send,
+	B: FnOnce() -> P,
```
Comment (bkchr) [inline collator/src/lib.rs]: Don't you need this to be `Send`?

File: collator/src/lib.rs [hunk 2]
```diff
@@ -461,51 +512,52 @@ fn compute_targets(para_id: ParaId, session_keys: &[ValidatorId], roster: DutyRo
 /// build by the given `BuildParachainContext` and arguments to the underlying polkadot node.
 ///
 /// This function blocks until done.
-pub fn run_collator<P>(
-	build_parachain_context: P,
+pub fn run_collator<P, B>(
```
Comment (bkchr) [inline collator/src/lib.rs]: Could this function not use `build_collator` internally?

File: collator/src/lib.rs (hunk 1 above)
Comment (cecton) [reply collator/src/lib.rs]: It compiles in cumulus so I guess not. I think it helps that this is an FnOnce and not a Fn. For substrate it was a Fn and I believe it would have worked better with an FnOnce. (but in substrate it doesn't make sense to put an FnOnce there)

File: collator/src/lib.rs (hunk 2 above)
Comment (cecton) [reply collator/src/lib.rs]: I tried but it doesn't work because build_collator is actually an async fn and I need a service to use `run_service_until_exit`. I could use build_collator and use run_until_exit (which takes a future instead) but then we would like some features like: setting the task_executor in the configuration and also the informant

File: collator/src/lib.rs [hunk 3]
```diff
@@ -447,6 +450,54 @@ fn run_collator_node<S, P, Extrinsic>(
 	Ok(service)
 }
 
+/// Build a collator node with the given `RelayChainContext` and `ParachainContext`
+/// build by the given `BuildParachainContext` and arguments to the underlying polkadot node.
+///
+/// This function does not block. It returns an `AbstractService` that can be used as a future
+pub async fn build_collator<P, B>(
```
Comment (cecton) [inline collator/src/lib.rs]: @bkchr maybe it would be better named start_collator in the end since it awaits for it

Comment (cecton) [reply collator/src/lib.rs]: I've done

File: collator/src/lib.rs [hunk 4]
```diff
@@ -345,6 +349,8 @@ fn run_collator_node<S, P, Extrinsic>(
 		spawner.clone(),
 	));
 
+	let build_parachain_context = build_parachain();
```
Comment (rphmeier) [inline collator/src/lib.rs]: why is it better to do this lazily?

File: collator/src/lib.rs [hunk 5]
```diff
@@ -273,12 +273,15 @@ impl<P: 'static, SP: 'static> RelayChainContext for ApiContext<P, SP> where
...
+/// Build a collator service with the given `RelayChainContext` and `ParachainContext`
+/// build by the given `BuildParachainContext` and arguments to the underlying polkadot node.
+///
+/// This function does not block. It returns an `AbstractService` that can be used as a future
+pub fn build_collator_service<S, P, Extrinsic, B>(
 	service: S,
 	para_id: ParaId,
 	key: Arc<CollatorPair>,
-	build_parachain_context: P,
+	build_parachain: B,
```
Comment (rphmeier) [inline collator/src/lib.rs]: maybe more readable to just do `build_parachain: impl FnOnce() -> P`

File: collator/src/lib.rs [hunk 6]
```diff
@@ -447,6 +453,52 @@ fn run_collator_node<S, P, Extrinsic>(
 	Ok(service)
 }
 
+/// Async function that will run the collator node with the given `RelayChainContext` and `ParachainContext`
+/// build by the given `BuildParachainContext` and arguments to the underlying polkadot node.
```
Comment (rphmeier) [inline collator/src/lib.rs]:
```suggestion
/// built by the given `BuildParachainContext` and arguments to the underlying polkadot node.
```

File: collator/src/lib.rs (hunk 4 above)
Comment (cecton) [reply collator/src/lib.rs]: It's not better. It's no choice. Polkadot has 4 different types of service and you can't return a common type in Rust

File: collator/src/lib.rs (hunk 5 above)
Comment (cecton) [reply collator/src/lib.rs]: I forgot about this syntax. Thanks!

File: collator/src/lib.rs (hunk 6 above)
Comment (cecton) [reply collator/src/lib.rs]: wow I copy-pasted this typo 3 times from the original comment without seeing it, thanks!

File: collator/src/lib.rs (hunk 4 above)
Comment (rphmeier) [reply collator/src/lib.rs]: I don't follow. I see that you are branching based on the type of service in `start_collator`, but the `build_parachain` parameter does not vary. So why does it need to be lazy?

Comment (cecton) [reply collator/src/lib.rs]: Yes I just realized this night I was telling crap :confounded: 

The reason for this was the hope to replace the CollatorBuilder. See [here](https://github.com/paritytech/cumulus/issues/44#issuecomment-583317236) for the discussion. Using a ParachainContext directly would be great but unfortunately I couldn't do it because of these 4 different types of service and the limitation.

I can totally remove the "builder" closure and keep only the CollatorBuilder *OR* I can remove the CollatorBuilder entirely and keep only the closure. cc @bkchr ? (But I must keep one: I cannot remove the need of a builder of some sort. Otherwise I need to duplicate some code like I did [here](https://github.com/paritytech/cumulus/pull/53#discussion_r376309209))

In my opinion, at this point, I think it would simplify a bit to remove the collator builder as this is just a fancy closure in the end.

Comment (cecton) [reply collator/src/lib.rs]: (Actually I couldn't figure out how to use only a closure so I will keep the BuildParachainContext/CollatorBuilder for now)

**Comment by cecton** [comment]
@bkchr @rphmeier this is ready for review again

File: collator/src/lib.rs [hunk 7]
```diff
@@ -273,8 +273,11 @@ impl<P: 'static, SP: 'static> RelayChainContext for ApiContext<P, SP> where
 	}
 }
 
-/// Run the collator node using the given `service`.
-fn run_collator_node<S, P, Extrinsic>(
+/// Build a collator service with the given `RelayChainContext` and `ParachainContext`
+/// built by the given `BuildParachainContext` and arguments to the underlying polkadot node.
+///
+/// This function does not block. It returns an `AbstractService` that can be used as a future
+pub fn build_collator_service<S, P, Extrinsic>(
```
Comment (bkchr) [inline collator/src/lib.rs]: Why is that required to be public?

Comment (cecton) [reply collator/src/lib.rs]: I thought it was used by the cli but actually it is not.
